src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,474 @@
+            paymentNonce[msg.sender] = nonce + 1;
+        }
+    }
+
+    /**
+     * @notice Claim payments against a given root (read from distributionRoots[claim.rootIndex]).
+     * Earnings are cumulative so earners don't have to claim against all distribution roots they have earnings for,
+     * they can simply claim against the latest root and the contract will calculate the difference between
+     * their cumulativeEarnings and cumulativeClaimed. This difference is then transferred to claimerFor[claim.earner]
+     * @param claim The PaymentMerkleClaim to be processed.
+     * Contains the root index, earner, payment leaves, and required proofs
+     * @dev only callable by the valid claimer, that is
+     * if claimerFor[claim.earner] is address(0) then only the earner can claim, otherwise only
+     * claimerFor[claim.earner] can claim the payments.
+     */
+    function processClaim(PaymentMerkleClaim calldata claim) external onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) nonReentrant {
+        DistributionRoot memory root = distributionRoots[claim.rootIndex];
+        require(_checkClaim(claim, root), "PaymentCoordinator.processClaim: claim does not pass the check");
+        // If claimerFor earner is not set, claimer is by default the earner. Else set to claimerFor
+        address earner = claim.earnerLeaf.earner;
+        address claimer = claimerFor[earner];
+        if (claimer == address(0)) {
+            claimer = earner;
+        }
+        require(msg.sender == claimer, "PaymentCoordinator.processClaim: caller is not valid claimer");
+        for (uint256 i = 0; i < claim.tokenIndices.length; i++) {
+            TokenTreeMerkleLeaf calldata tokenLeaf = claim.tokenLeaves[i];
+
+            // Calculate amount to claim and update cumulativeClaimed
+            // Will revert if new leaf cumulativeEarnings is less than cumulative claimed
+            uint256 claimAmount = tokenLeaf.cumulativeEarnings - cumulativeClaimed[earner][tokenLeaf.token];
+            cumulativeClaimed[earner][tokenLeaf.token] = tokenLeaf.cumulativeEarnings;
+
+            tokenLeaf.token.safeTransfer(claimer, claimAmount);
+            emit PaymentClaimed(root.root, earner, claimer, claim.tokenLeaves[i].token, claimAmount);
+        }
+    }
+
+    /**
+     * @notice Creates a new distribution root. activatedAt is set to block.timestamp + activationDelay
+     * @param root The merkle root of the distribution
+     * @param paymentCalculationEndTimestamp The timestamp until which payments have been calculated
+     * @dev Only callable by the paymentUpdater
+     */
+    function submitRoot(
+        bytes32 root,
+        uint32 paymentCalculationEndTimestamp
+    ) external onlyPaymentUpdater {
+        require(
+            paymentCalculationEndTimestamp > currPaymentCalculationEndTimestamp,
+            "PaymentCoordinator.submitRoot: new root must be for newer calculated period"
+        );
```

Comment:
What is this used for? It doesn't seem to be a particularly useful check - there are no rules enforcing this field's value except that it's larger than the last time `submitRoot` was called.

If (for whatever reason) the payment updater calls this with `type(uint32).max`, `submitRoot` is bricked until we upgrade and fix it.

Is there some upper bound on how much this can increase by, or some relation it should have to `block.timestamp`?

Reply:
This check is just a sanity check for the paymentUpdater to submit roots sequentially and `currPaymentCalculationEndTimestamp` is read from the backend to know where to start calculating payments from.
We have MAX_RETROACTIVE_LENGTH and MAX_FUTURE_LENGTH for valid startTimestamps for RangePayments.
I think we can just reuse MAX_FUTURE_LENGTH to check @gpsanant 
```
        require(
            paymentCalculationEndTimestamp < block.timestamp + MAX_FUTURE_LENGTH,
            "PaymentCoordinator.submitRoot: paymentCalculationEndTimestamp too far in the future"
        );
```

Reply:
Will go with this since it doesn't make sense for payments to be calculated in the future
```
        require(
            paymentCalculationEndTimestamp < block.timestamp,
            "PaymentCoordinator.submitRoot: paymentCalculationEndTimestamp can't be in the future"
        );
```